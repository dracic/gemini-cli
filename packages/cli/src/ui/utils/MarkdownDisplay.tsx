/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * MarkdownDisplay Component - AST-Based Rendering
 *
 * Simplified component that delegates to AstToInkTransformer for all markdown parsing.
 * Replaces regex-based line-by-line parsing with semantic AST transformation.
 *
 * Key improvements:
 * - Consistent spacing regardless of source markdown formatting
 * - Proper nested structure handling (lists, code blocks)
 * - Graceful fallback for malformed markdown
 */

import React from 'react';
import { Text } from 'ink';
import { transformMarkdownToInk } from './AstToInkTransformer.js';

interface MarkdownDisplayProps {
  text: string;
  isPending: boolean;
  availableTerminalHeight?: number;
  terminalWidth: number;
}

const MarkdownDisplayInternal: React.FC<MarkdownDisplayProps> = ({
  text,
  isPending,
  availableTerminalHeight,
  terminalWidth,
}) => {
  // Empty text check
  if (!text) return <></>;

  // Delegate to AST transformer
  const rendered = transformMarkdownToInk(text, {
    isPending,
    availableTerminalHeight,
    terminalWidth,
  });

  // Fallback for unexpected transformer output
  if (!rendered) {
    return <Text>{text}</Text>;
  }

  return <>{rendered}</>;
};

export const MarkdownDisplay = React.memo(MarkdownDisplayInternal);
