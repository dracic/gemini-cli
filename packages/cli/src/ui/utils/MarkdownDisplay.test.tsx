/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { MarkdownDisplay } from './MarkdownDisplay.js';
import { LoadedSettings } from '../../config/settings.js';
import { SettingsContext } from '../contexts/SettingsContext.js';

describe('<MarkdownDisplay />', () => {
  const baseProps = {
    isPending: false,
    terminalWidth: 80,
    availableTerminalHeight: 40,
  };

  const mockSettings = new LoadedSettings(
    { path: '', settings: {}, originalSettings: {} },
    { path: '', settings: {}, originalSettings: {} },
    { path: '', settings: {}, originalSettings: {} },
    { path: '', settings: {}, originalSettings: {} },
    true,
    new Set(),
  );

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders nothing for empty text', () => {
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text="" />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a simple paragraph', () => {
    const text = 'Hello, world.';
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders headers with correct levels', () => {
    const text = `
# Header 1
## Header 2
### Header 3
#### Header 4
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a fenced code block with a language', () => {
    const text = '```javascript\nconst x = 1;\nconsole.log(x);\n```';
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a fenced code block without a language', () => {
    const text = '```\nplain text\n```';
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('handles unclosed (pending) code blocks', () => {
    const text = '```typescript\nlet y = 2;';
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} isPending={true} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders unordered lists with different markers', () => {
    const text = `
- item A
* item B
+ item C
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders nested unordered lists', () => {
    const text = `
* Level 1
  * Level 2
    * Level 3
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders ordered lists', () => {
    const text = `
1. First item
2. Second item
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders horizontal rules', () => {
    const text = `
Hello
---
World
***
Test
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders tables correctly', () => {
    const text = `
| Header 1 | Header 2 |
|----------|:--------:|
| Cell 1   | Cell 2   |
| Cell 3   | Cell 4   |
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('handles a table at the end of the input', () => {
    const text = `
Some text before.
| A | B |
|---|
| 1 | 2 |`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('inserts a single space between paragraphs', () => {
    const text = `Paragraph 1.

Paragraph 2.`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('correctly parses a mix of markdown elements', () => {
    const text = `
# Main Title

Here is a paragraph.

- List item 1
- List item 2

\`\`\`
some code
\`\`\`

Another paragraph.
`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('hides line numbers in code blocks when showLineNumbers is false', () => {
    const text = '```javascript\nconst x = 1;\n```';
    const settings = new LoadedSettings(
      { path: '', settings: {}, originalSettings: {} },
      { path: '', settings: {}, originalSettings: {} },
      {
        path: '',
        settings: { ui: { showLineNumbers: false } },
        originalSettings: { ui: { showLineNumbers: false } },
      },
      { path: '', settings: {}, originalSettings: {} },
      true,
      new Set(),
    );

    const { lastFrame } = render(
      <SettingsContext.Provider value={settings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
    expect(lastFrame()).not.toContain(' 1 ');
  });

  it('shows line numbers in code blocks by default', () => {
    const text = '```javascript\nconst x = 1;\n```';
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay {...baseProps} text={text} />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
    expect(lastFrame()).toContain(' 1 ');
  });

  it('correctly renders a file with Unix (LF) line endings', () => {
    const markdownContent = `# Programming Languages Overview

## 1. Python

- **Strengths:**
  - Easy to learn syntax
  - Extensive library ecosystem
  - Great for data science
    - NumPy for numerical computing
    - Pandas for data analysis
    - Matplotlib for visualization
- **Use Cases:**
  - Web development with Django/Flask
  - Machine learning and AI
  - Automation and scripting

## 2. JavaScript

- **Strengths:**
  - Universal language (frontend/backend)`;
    const { lastFrame } = render(
      <SettingsContext.Provider value={mockSettings}>
        <MarkdownDisplay
          text={markdownContent}
          isPending={false}
          availableTerminalHeight={100} // Provide ample height
          terminalWidth={120} // Provide ample width
        />
      </SettingsContext.Provider>,
    );
    expect(lastFrame()).toMatchSnapshot();
  });
});
