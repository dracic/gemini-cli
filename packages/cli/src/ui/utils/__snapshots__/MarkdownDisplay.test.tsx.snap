// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MarkdownDisplay /> > handles malformed markdown gracefully 1`] = `
" 1 Unclosed code block
"
`;

exports[`<MarkdownDisplay /> > renders a simple paragraph 1`] = `
"Hello, world.
"
`;

exports[`<MarkdownDisplay /> > renders large 2KB markdown document 1`] = `
"Section 1

This is paragraph 1 with some bold and italic text.

Section 2

This is paragraph 2 with some bold and italic text.

Section 3

This is paragraph 3 with some bold and italic text.

Section 4

This is paragraph 4 with some bold and italic text.

Section 5

This is paragraph 5 with some bold and italic text.

Section 6

This is paragraph 6 with some bold and italic text.

Section 7

This is paragraph 7 with some bold and italic text.

Section 8

This is paragraph 8 with some bold and italic text.

Section 9

This is paragraph 9 with some bold and italic text.

Section 10

This is paragraph 10 with some bold and italic text.

Section 11

This is paragraph 11 with some bold and italic text.

Section 12

This is paragraph 12 with some bold and italic text.

Section 13

This is paragraph 13 with some bold and italic text.

Section 14

This is paragraph 14 with some bold and italic text.

Section 15

This is paragraph 15 with some bold and italic text.

Section 16

This is paragraph 16 with some bold and italic text.

Section 17

This is paragraph 17 with some bold and italic text.

Section 18

This is paragraph 18 with some bold and italic text.

Section 19

This is paragraph 19 with some bold and italic text.

Section 20

This is paragraph 20 with some bold and italic text.

Section 21

This is paragraph 21 with some bold and italic text.

Section 22

This is paragraph 22 with some bold and italic text.

Section 23

This is paragraph 23 with some bold and italic text.

Section 24

This is paragraph 24 with some bold and italic text.

Section 25

This is paragraph 25 with some bold and italic text.

Section 26

This is paragraph 26 with some bold and italic text.

Section 27

This is paragraph 27 with some bold and italic text.

Section 28

This is paragraph 28 with some bold and italic text.
"
`;

exports[`<MarkdownDisplay /> > renders nothing for empty text 1`] = `""`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > correctly parses a mix of markdown elements 1`] = `
"Main Title

Here is a paragraph.

* List item 1
* List item 2

 1 some code

Another paragraph.
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > handles a table at the end of the input 1`] = `
"Some text before.
| A | B |
|---|
| 1 | 2 |
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > handles deeply nested lists (10 levels) 1`] = `
"*    *
L1L2  * L3  * L4     * L5      * L6         * L7          * L8           * L9             * L10

"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > handles unclosed (pending) code blocks 1`] = `
" 1 let y = 2;
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > hides line numbers in code blocks when showLineNumbers is false 1`] = `
" const x = 1;
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > inserts a single space between paragraphs 1`] = `
"Paragraph 1.

Paragraph 2.
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders 4-level nested lists 1`] = `
"* Level 1   * Level 2      * Level 3         * Level 4
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders a fenced code block with a language 1`] = `
" 1 const x = 1;
 2 console.log(x);
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders a fenced code block without a language 1`] = `
" 1 plain text
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders blockquotes 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ This is a quote                                                                                  │
│ It spans multiple lines                                                                          │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders code blocks within list items 1`] = `
"* List item with         1 const x = 1;
code:
* Another list item
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders headers with correct levels 1`] = `
"Header 1

Header 2

Header 3

Header 4
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders horizontal rules 1`] = `
"Hello

World

---

Test
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders inline code within text 1`] = `
"Use the \`console.log()\` function for debugging.
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders mixed ordered and unordered list nesting 1`] = `
"1. Ordered item 1   * Unordered sub-item
                    * Another sub-item
2. Ordered item 2   1. Nested ordered
                    2. Another nested
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders nested blockquotes 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Level 1 quote┌──────────────────────────────┐                                                    │
│              │ Level 2 quote┌──────────────┐│                                                    │
│              │              │ Level 3 quote││                                                    │
│              │              └──────────────┘│                                                    │
│              └──────────────────────────────┘                                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders nested inline elements (bold within link) 1`] = `
"bold link
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders nested unordered lists 1`] = `
"* Level 1   * Level 2      * Level 3
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders ordered list with custom start property 1`] = `
"5. Item five
6. Item six
7. Item seven
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders ordered lists 1`] = `
"1. First item
2. Second item
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders strikethrough text 1`] = `
"This is deleted text.
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders tables correctly 1`] = `
"
┌──────────┬──────────┐
│ Header 1 │ Header 2 │
├──────────┼──────────┤
│ Cell 1   │ Cell 2   │
│ Cell 3   │ Cell 4   │
└──────────┴──────────┘

"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > renders unordered lists with different markers 1`] = `
"* item A

* item B

* item C
"
`;

exports[`<MarkdownDisplay /> > with 'Unix' line endings > shows line numbers in code blocks by default 1`] = `
" 1 const x = 1;
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > correctly parses a mix of markdown elements 1`] = `
"Main Title

Here is a paragraph.

* List item 1
* List item 2

 1 some code

Another paragraph.
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > handles a table at the end of the input 1`] = `
"Some text before.
| A | B |
|---|
| 1 | 2 |
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > handles deeply nested lists (10 levels) 1`] = `
"*    *
L1L2  * L3  * L4     * L5      * L6         * L7          * L8           * L9             * L10

"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > handles unclosed (pending) code blocks 1`] = `
" 1 let y = 2;
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > hides line numbers in code blocks when showLineNumbers is false 1`] = `
" const x = 1;
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > inserts a single space between paragraphs 1`] = `
"Paragraph 1.

Paragraph 2.
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders 4-level nested lists 1`] = `
"* Level 1   * Level 2      * Level 3         * Level 4
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders a fenced code block with a language 1`] = `
" 1 const x = 1;
 2 console.log(x);
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders a fenced code block without a language 1`] = `
" 1 plain text
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders blockquotes 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ This is a quote
                                                                                 │
│ It spans multiple lines                                                                          │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders code blocks within list items 1`] = `
"* List item with         1 const x = 1;
code:
* Another list item
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders headers with correct levels 1`] = `
"Header 1

Header 2

Header 3

Header 4
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders horizontal rules 1`] = `
"Hello

World

---

Test
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders inline code within text 1`] = `
"Use the \`console.log()\` function for debugging.
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders mixed ordered and unordered list nesting 1`] = `
"1. Ordered item 1   * Unordered sub-item
                    * Another sub-item
2. Ordered item 2   1. Nested ordered
                    2. Another nested
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders nested blockquotes 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Level 1 quote┌──────────────────────────────┐                                                    │
│              │ Level 2 quote┌──────────────┐│                                                    │
│              │              │ Level 3 quote││                                                    │
│              │              └──────────────┘│                                                    │
│              └──────────────────────────────┘                                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders nested inline elements (bold within link) 1`] = `
"bold link
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders nested unordered lists 1`] = `
"* Level 1   * Level 2      * Level 3
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders ordered list with custom start property 1`] = `
"5. Item five
6. Item six
7. Item seven
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders ordered lists 1`] = `
"1. First item
2. Second item
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders strikethrough text 1`] = `
"This is deleted text.
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders tables correctly 1`] = `
"
┌──────────┬──────────┐
│ Header 1 │ Header 2 │
├──────────┼──────────┤
│ Cell 1   │ Cell 2   │
│ Cell 3   │ Cell 4   │
└──────────┴──────────┘

"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > renders unordered lists with different markers 1`] = `
"* item A

* item B

* item C
"
`;

exports[`<MarkdownDisplay /> > with 'Windows' line endings > shows line numbers in code blocks by default 1`] = `
" 1 const x = 1;
"
`;
